<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px"> 

## Lab: SQL Practice

-----

### Required Packages

To run SQL queries in this Jupyter notebook, install the required packages using these commands:

```bash
conda install sqlalchemy
conda install psycopg2
```

If you encounter an error while installing [**psycopg2**](https://pypi.org/project/psycopg2/), you can install the stand-alone package, which is sufficient for completing this lab, by running the following command in your terminal:

```bash 
pip install psycopg2-binary
```

If pip doesn't work for you, try... 
```bash
pip3 install psycopg2-binary
```

### Instructions

This lab consists of ten prompts. For each prompt, you will need to:
- Write a SQL query
- Use the query results to answer the question(s)

The purpose of this lab is to develop your SQL proficiency - a critical skill in data science. While these problems could be solved using Pandas, your goal should be to complete all exercises using **_only_** SQL.

**Data**: This lab uses the Iowa liquor database, which contains multiple tables tracking liquor transactions. 

A database [**schema**](https://hasura.io/learn/database/postgresql/core-concepts/1-postgresql-schema/#:~:text=Schema%20is%20a%20collection%20of,different%20features%20into%20different%20schemas.) is a diagram that shows the structure and organization of a database, including its tables, columns, and the relationships between them. 

The database schema is shown below:

<img src="./images/schema.png" alt="schema" width="750"/>

In [4]:
# Import pandas library for data manipulation and analysis
# pandas provides powerful DataFrame objects and data manipulation tools
import pandas as pd

# Import create_engine from SQLAlchemy - this is the main factory function 
# used to create a new database engine that handles connections and queries
from sqlalchemy import create_engine

# Create a database engine that will manage connections to the database
# The engine is the starting point for any SQLAlchemy application
# Format: create_engine('dialect+driver://username:password@host:port/database')
engine = \
create_engine('postgresql://analytics_student:analyticsga@analyticsga-psql.generalassemb.ly:5432/iowa_liquor_sales_database')

### Prompt 1

How many rows are in the **stores** table of the Iowa liquor database?

In [6]:
# Use a SQL query to find the answer to the above prompt

sql = """
SELECT COUNT(*) FROM stores
"""

pd.read_sql_query(sql, engine)

count
0   1973

**Prompt 1 Answer**:

1973

### Prompt 2

When sorting stores alphabetically by **store_name**, which store appears fourth in the list?

In [10]:
# Use a SQL query to find the answer to the above prompt

sql = """
SELECT name
FROM stores
ORDER BY name
LIMIT 1 OFFSET 3
"""

pd.read_sql_query(sql, engine)

name
0  3061 Cub Foods / Sioux City

**Prompt 2 Answer**:

3061 Cub Foods / Sioux City

### Prompt 3

Among those sales whose category name is **IMPORTED VODKA**, if you sort by vendor in alphabetical order, what is the bottle price and number of bottles bought in the first 3 transactions?
> You should give us three sets of numbers - one for each transaction.

In [14]:
# Use a SQL query to find the answer to the above prompt

sql = """
SELECT p.bottle_price, COUNT(s.item) AS bottle_number
FROM sales s
JOIN products p
ON s.item=p.item_no
WHERE s.category_name ILIKE '%%IMPORTED VODKA%%'
GROUP BY p.item_no
ORDER BY p.vendor
LIMIT 3
"""

pd.read_sql_query(sql, engine)

bottle_price  bottle_number
0       $19.00             34
1        $5.99              1
2       $18.49            903

**Prompt 3 Answer**: 


`id`	`bottle_price`	`bottle_number`
```
0	$19.00	34
1	$5.99	1
2	$5.00	435


### Prompt 4

What was the total amount of liquor sold during February 2015?

In [18]:
# Use a SQL query to find the answer to the above prompt

sql = """
SELECT SUM(s.total)
FROM sales s
WHERE date BETWEEN '2015-02-01' AND '2015-02-28'
"""
pd.read_sql_query(sql, engine)

sum
0  21295350.55

**Prompt 4 Answer**:

21,295,350.55

### Prompt 5

For all transactions involving **IMPORTED VODKA**, find the vendor with the most transactions
- Report their:
  - Total transaction count
  - Total bottles sold
  - Total revenue

Use clear, descriptive column names in your results.

In [22]:
# Use a SQL query to find the answer to the above prompt

sql = """
SELECT p.vendor_name, COUNT(*) AS total_transaction_count, 
    SUM(s.bottle_qty) AS total_bottle_sold, 
    SUM(s.total) AS total_revenue
FROM sales s
JOIN products p
ON s.item=p.item_no
WHERE s.category_name ILIKE '%%IMPORTED VODKA%%'
GROUP BY p.vendor_name
ORDER BY COUNT(*) DESC
LIMIT 1
"""

pd.read_sql_query(sql, engine).T

0
vendor_name              Pernod Ricard Usa/austin Nichols
total_transaction_count                             67190
total_bottle_sold                                  570679
total_revenue                                  10902960.4

**Prompt 5 Answer**: 

<pre>
vendor_name	Pernod Ricard Usa/austin Nichols
total_transaction_count	67190
total_bottle_sold	570679
total_revenue	10902960.4.88
</pre>

### Prompt 6

In retail, **markup** represents the difference between a product's selling price and its cost - essentially the profit per item.
For Iowa liquor sales, state law sets minimum prices. 

Let's calculate the **markup** (the difference between **btl_price** and **state_btl_cost**).

> If you haven't subtracted columns together before, [this StackOverflow question](https://stackoverflow.com/questions/7536996/multiplying-two-columns-in-sql-server/7537059) may be helpful.

In [26]:
# Use a SQL query to find the answer to the above prompt

sql = """
SELECT btl_price - state_btl_cost AS markup
FROM sales
LIMIT 5
"""

pd.read_sql_query(sql, engine)

markup
0  $5.75
1  $5.75
2  $5.75
3  $8.58
4  $2.50

**Prompt 6 Answer**: 

<pre>
	markup
0	$5.75
1	$5.75
2	$5.75
3	$8.58
4	$2.50
</pre>

### Prompt 7

Calculate the **Percentage Markup** by dividing markup by the state bottle cost to determine how much each store increases prices above the state minimum.
- _Rather than the total profit per bottle, this will tell you how much stores are increasing their price over the state minimum._

After calculating the **Percentage Markup**, calculate the average percentage markup per **store**. 

Then find the store with the highest average markup percentage. What is this store's ID and average markup?

In [30]:
# Use a SQL query to find the answer to the above prompt

sql = """
SELECT st.store AS store_id, AVG(percentage_markup) AS avg_percentage_markup
FROM (SELECT store, (btl_price - state_btl_cost) / state_btl_cost * 100 AS percentage_markup
      FROM sales) sm
JOIN stores st
ON sm.store=st.store
GROUP BY st.store
ORDER BY avg_percentage_markup DESC
LIMIT 1
"""

pd.read_sql_query(sql, engine)

store_id  avg_percentage_markup
0      4013              52.547124

**Prompt 7 Answer**: 

<pre>
	store_id	avg_percentage_markup
0	4013	52.547124
</pre>

### Prompt 8

Building off of your last query, identify the five locations with the highest average markup percentage. For each, provide:
- Store ID
- Average, minimum, and maximum markup percentages
- Name and address

What are the names of the five stores?

In [34]:
# Use a SQL query to find the answer to the above prompt

sql = """
SELECT st.store AS store_id, 
    AVG(percentage_markup) AS avg_percentage_markup,
    MIN(percentage_markup) AS min_percentage_markup,
    MAX(percentage_markup) AS max_percentage_markup
FROM (SELECT store, (btl_price - state_btl_cost) / state_btl_cost * 100 AS percentage_markup
      FROM sales) sm
JOIN stores st
ON sm.store=st.store
GROUP BY st.store
ORDER BY avg_percentage_markup DESC
LIMIT 5
"""

pd.read_sql_query(sql, engine)

store_id  avg_percentage_markup  min_percentage_markup  \
0      4013              52.547124              49.857143   
1      4024              52.149880              49.846154   
2      4266              51.890342              49.846154   
3      3833              51.509378              49.846154   
4      3660              51.459084              49.846154   

   max_percentage_markup  
0             670.876289  
1             670.876289  
2             670.876289  
3             670.876289  
4             670.876289

**Prompt 8 Answer**:

<pre>
store_id	avg_percentage_markup	min_percentage_markup	max_percentage_markup
0	4013	52.547124	49.857143	670.876289
1	4024	52.149880	49.846154	670.876289
2	4266	51.890342	49.846154	670.876289
3	3833	51.509378	49.846154	670.876289
4	3660	51.459084	49.846154	670.87628976289
</pre>

### Prompt 9

Which nine (9) counties sell the highest number of bottles of liquor per capita? 

For these counties:
- Calculate the average bottle size (rounded to one decimal place)
- Identify which county sells the largest average bottle size

**_BONUS CHALLENGE_**: Try solving this using a [**SQL subquery**](https://www.dofactory.com/sql/subquery) – a query within a query!

In [38]:
# Use a SQL query to find the answer to the above prompt

sql = """
SELECT cs.county, 
cs.bottle_number_per_capita, 
ROUND(cs.avg_bottle_size, 1) AS avg_bottle_size
FROM (SELECT sa.county, SUM(sa.bottle_qty) / co.population AS bottle_number_per_capita, 
      AVG(pr.bottle_size) AS avg_bottle_size
        FROM sales sa
        JOIN products pr
        ON sa.item=pr.item_no
        JOIN counties co
        ON sa.county=co.county
        GROUP BY sa.county, co.population
        ORDER BY bottle_number_per_capita DESC
        LIMIT 9) cs
ORDER BY bottle_number_per_capita DESC
"""

pd.read_sql_query(sql, engine)

county  bottle_number_per_capita  avg_bottle_size
0      Dickinson                        22            974.1
1           Polk                        15            892.6
2     Black Hawk                        15            862.9
3    Cerro Gordo                        14            986.4
4          Scott                        13            871.6
5        Johnson                        13            895.4
6           Linn                        12            886.3
7  Pottawattamie                        11            893.7
8        Kossuth                        11           1041.0

**Prompt 9 Answer**:

- Kossuth	sells the largest average bottle size
<pre>
	county		bottle_number_per_capita	avg_bottle_size
0	Dickinson	22				974.1
1	Polk		15				892.6
2	Black Hawk	15				862.9
3	Cerro Gordo	14				986.4
4	Scott		13				871.6
5	Johnson		13				895.4
6	Linn		12				886.3
7	Pottawattamie	11				893.7
8	Kossuth		11				1041.0




</pre>

### Prompt 10
 
In the **sales** table, examine the relationship between three columns:
- **btl_price**
- **bottle_qty**
- **total** 

The total column should equal **btl_price** multiplied by **bottle_qty**. 

How many transactions have a value of **total** that is _not equal_ to **btl_price** times **bottle_qty**?

Note: You'll need to convert the **total** column to type [**money**](https://www.postgresql.org/docs/current/datatype-money.html) for proper comparison.

In [42]:
# Use a SQL query to find the answer to the above prompt

sql = """
SELECT COUNT(*) AS num
FROM sales
WHERE total::money != (btl_price * bottle_qty)
"""

pd.read_sql_query(sql, engine)

num
0    0

**Prompt 10 Answer**:

- None